File: backend/app/utils/validation.py
import re
from typing import Optional
from datetime import datetime

# Common password denylist
COMMON_PASSWORDS = {
    "password", "1234567890", "qwerty", "qwerty123", "password123",
    "12345678", "123456789", "abc123", "password1", "welcome123"
}


def validate_email(email: str) -> tuple[bool, Optional[str]]:
    """Validate email format. Returns (is_valid, error_message)."""
    if not email:
        return False, "Email is required"
    
    email = email.strip().lower()
    
    # Basic email regex
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    if not re.match(pattern, email):
        return False, "Please enter a valid email address"
    
    if len(email) > 254:  # RFC 5321 limit
        return False, "Email address is too long"
    
    return True, None


def validate_name(name: str) -> tuple[bool, Optional[str]]:
    """Validate name. Returns (is_valid, error_message)."""
    if not name:
        return False, "Name is required"
    
    name = name.strip()
    
    if len(name) < 2:
        return False, "Name must be at least 2 characters"
    
    if len(name) > 40:
        return False, "Name must be no more than 40 characters"
    
    # Allow letters, spaces, hyphens, apostrophes
    pattern = r'^[a-zA-Z\s\-\']+$'
    if not re.match(pattern, name):
        return False, "Name can only contain letters, spaces, hyphens, and apostrophes"
    
    if name != name.strip():
        return False, "Name cannot have leading or trailing spaces"
    
    return True, None


def validate_school(school: str) -> tuple[bool, Optional[str]]:
    """Validate school name. Returns (is_valid, error_message)."""
    if not school:
        return False, "School is required"
    
    school = school.strip()
    
    if len(school) < 2:
        return False, "School name must be at least 2 characters"
    
    if len(school) > 60:
        return False, "School name must be no more than 60 characters"
    
    # Safe characters: letters, numbers, spaces, dots, hyphens, apostrophes
    pattern = r'^[a-zA-Z0-9\s\.\-\']+$'
    if not re.match(pattern, school):
        return False, "School name contains invalid characters"
    
    return True, None


def validate_grad_year(grad_year: int) -> tuple[bool, Optional[str]]:
    """Validate graduation year. Returns (is_valid, error_message)."""
    current_year = datetime.now().year
    min_year = current_year - 10
    max_year = current_year + 10
    
    if grad_year < min_year or grad_year > max_year:
        return False, f"Graduation year must be between {min_year} and {max_year}"
    
    return True, None


def validate_monthly_goal(amount: float) -> tuple[bool, Optional[str]]:
    """Validate monthly goal amount. Returns (is_valid, error_message)."""
    if amount < 0:
        return False, "Monthly goal must be positive"
    
    if amount > 5000:
        return False, "Monthly goal cannot exceed $5,000"
    
    return True, None


def validate_password(password: str) -> tuple[bool, Optional[str]]:
    """Validate password strength. Returns (is_valid, error_message)."""
    if not password:
        return False, "Password is required"
    
    if len(password) < 10:
        return False, "Password must be at least 10 characters"
    
    if len(password) > 72:
        return False, "Password must be no more than 72 characters"
    
    # Check for uppercase
    if not re.search(r'[A-Z]', password):
        return False, "Password must include at least one uppercase letter"
    
    # Check for lowercase
    if not re.search(r'[a-z]', password):
        return False, "Password must include at least one lowercase letter"
    
    # Check for digit
    if not re.search(r'\d', password):
        return False, "Password must include at least one number"
    
    # Check for symbol
    if not re.search(r'[!@#$%^&*()_+\-=\[\]{};\':"\\|,.<>/?]', password):
        return False, "Password must include at least one symbol"
    
    # Check against common passwords
    if password.lower() in COMMON_PASSWORDS:
        return False, "This password is too common. Please choose a stronger password"
    
    return True, None

File: ios/Piggie/Core/Validation.swift
import Foundation

struct ValidationResult {
    let isValid: Bool
    let errorMessage: String?
}

class Validator {
    static let shared = Validator()
    
    private init() {}
    
    func validateEmail(_ email: String) -> ValidationResult {
        if email.isEmpty {
            return ValidationResult(isValid: false, errorMessage: "Email is required")
        }
        
        let trimmed = email.trimmingCharacters(in: .whitespaces).lowercased()
        let pattern = #"^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$"#
        let regex = try! NSRegularExpression(pattern: pattern)
        let range = NSRange(location: 0, length: trimmed.utf16.count)
        
        if regex.firstMatch(in: trimmed, range: range) == nil {
            return ValidationResult(isValid: false, errorMessage: "Please enter a valid email address")
        }
        
        if trimmed.count > 254 {
            return ValidationResult(isValid: false, errorMessage: "Email address is too long")
        }
        
        return ValidationResult(isValid: true, errorMessage: nil)
    }
    
    func validateName(_ name: String) -> ValidationResult {
        if name.isEmpty {
            return ValidationResult(isValid: false, errorMessage: "Name is required")
        }
        
        let trimmed = name.trimmingCharacters(in: .whitespaces)
        
        if trimmed.count < 2 {
            return ValidationResult(isValid: false, errorMessage: "Name must be at least 2 characters")
        }
        
        if trimmed.count > 40 {
            return ValidationResult(isValid: false, errorMessage: "Name must be no more than 40 characters")
        }
        
        let pattern = #"^[a-zA-Z\s\-\']+$"#
        let regex = try! NSRegularExpression(pattern: pattern)
        let range = NSRange(location: 0, length: trimmed.utf16.count)
        
        if regex.firstMatch(in: trimmed, range: range) == nil {
            return ValidationResult(isValid: false, errorMessage: "Name can only contain letters, spaces, hyphens, and apostrophes")
        }
        
        return ValidationResult(isValid: true, errorMessage: nil)
    }
    
    func validateSchool(_ school: String) -> ValidationResult {
        if school.isEmpty {
            return ValidationResult(isValid: false, errorMessage: "School is required")
        }
        
        let trimmed = school.trimmingCharacters(in: .whitespaces)
        
        if trimmed.count < 2 {
            return ValidationResult(isValid: false, errorMessage: "School name must be at least 2 characters")
        }
        
        if trimmed.count > 60 {
            return ValidationResult(isValid: false, errorMessage: "School name must be no more than 60 characters")
        }
        
        return ValidationResult(isValid: true, errorMessage: nil)
    }
    
    func validateGradYear(_ year: Int) -> ValidationResult {
        let currentYear = Calendar.current.component(.year, from: Date())
        let minYear = currentYear - 10
        let maxYear = currentYear + 10
        
        if year < minYear || year > maxYear {
            return ValidationResult(isValid: false, errorMessage: "Graduation year must be between \(minYear) and \(maxYear)")
        }
        
        return ValidationResult(isValid: true, errorMessage: nil)
    }
    
    func validateMonthlyGoal(_ amount: Double) -> ValidationResult {
        if amount < 0 {
            return ValidationResult(isValid: false, errorMessage: "Monthly goal must be positive")
        }
        
        if amount > 5000 {
            return ValidationResult(isValid: false, errorMessage: "Monthly goal cannot exceed $5,000")
        }
        
        return ValidationResult(isValid: true, errorMessage: nil)
    }
    
    func validatePassword(_ password: String) -> ValidationResult {
        if password.isEmpty {
            return ValidationResult(isValid: false, errorMessage: "Password is required")
        }
        
        if password.count < 10 {
            return ValidationResult(isValid: false, errorMessage: "Password must be at least 10 characters")
        }
        
        if password.count > 72 {
            return ValidationResult(isValid: false, errorMessage: "Password must be no more than 72 characters")
        }
        
        let hasUpperCase = password.rangeOfCharacter(from: CharacterSet.uppercaseLetters) != nil
        let hasLowerCase = password.rangeOfCharacter(from: CharacterSet.lowercaseLetters) != nil
        let hasDigit = password.rangeOfCharacter(from: CharacterSet.decimalDigits) != nil
        let hasSymbol = password.rangeOfCharacter(from: CharacterSet(charactersIn: "!@#$%^&*()_+-=[]{}|;':\",./<>?")) != nil
        
        if !hasUpperCase {
            return ValidationResult(isValid: false, errorMessage: "Password must include at least one uppercase letter")
        }
        
        if !hasLowerCase {
            return ValidationResult(isValid: false, errorMessage: "Password must include at least one lowercase letter")
        }
        
        if !hasDigit {
            return ValidationResult(isValid: false, errorMessage: "Password must include at least one number")
        }
        
        if !hasSymbol {
            return ValidationResult(isValid: false, errorMessage: "Password must include at least one symbol")
        }
        
        let commonPasswords = ["password", "1234567890", "qwerty", "qwerty123", "password123"]
        if commonPasswords.contains(password.lowercased()) {
            return ValidationResult(isValid: false, errorMessage: "This password is too common. Please choose a stronger password")
        }
        
        return ValidationResult(isValid: true, errorMessage: nil)
    }
}
